modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeVelocityVariableReferenceMigrationCheck.java

```diff
@@ -0,0 +1,198 @@
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.source.formatter.check;
+
+import com.liferay.petra.string.CharPool;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.source.formatter.check.util.VelocityMigrationUtil;
+
+import java.io.IOException;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * @author Nícolas Moura
+ */
+public class UpgradeVelocityVariableReferenceMigrationCheck
+	extends BaseFileCheck {
+
+	@Override
+	protected String doProcess(
+			String fileName, String absolutePath, String content)
+		throws IOException {
+
+		if (!fileName.endsWith(".vm")) {
+			return content;
+		}
+
+		VelocityMigrationUtil.writeMigratedContent(
+			_formatContent(
+				VelocityMigrationUtil.getFreeMarkerMigratedContent(
+					content, fileName)),
+			fileName);
+
+		return content;
+	}
+
+	private static boolean _isAttribute(String line, String match) {
+		boolean attribute = false;
+
+		int initMatchIndex = line.indexOf(match);
+
+		String afterMatch = line.substring(initMatchIndex + match.length());
+		String beforeMatch = line.substring(0, initMatchIndex);
+
+		for (int i = 1; i < (beforeMatch.length() + 1); i++) {
+			int nextCharIndex = initMatchIndex - i;
+
+			if ((line.charAt(nextCharIndex) == CharPool.QUOTE) &&
+				((StringUtil.count(afterMatch, CharPool.QUOTE) % 2) != 0)) {
+
+				attribute = true;
+
+				break;
+			}
+
+			if ((line.charAt(nextCharIndex) == CharPool.APOSTROPHE) &&
+				((StringUtil.count(afterMatch, CharPool.APOSTROPHE) % 2) !=
+					0)) {
+
+				attribute = true;
+
+				break;
+			}
+		}
+
+		return attribute;
+	}
+
+	private String _formatContent(String content) {
+		String[] lines = content.split(StringPool.NEW_LINE);
+
+		for (String line : lines) {
+			Matcher matcher = _variableReferencePattern.matcher(line);
+
+			String newLine = line;
+
+			while (matcher.find()) {
+				String match = matcher.group();
+
+				if (_isJQuery(match)) {
+					continue;
+				}
+
+				if (!_isValidReplacement(newLine, match)) {
+					newLine = StringUtil.replace(
+						newLine, match,
+						StringUtil.removeChar(match, CharPool.DOLLAR));
+
+					continue;
+				}
+
+				int beginCharIndex = newLine.indexOf(match) + match.length();
+
+				if (newLine.charAt(beginCharIndex) ==
+						CharPool.OPEN_PARENTHESIS) {
+
+					String endLine = newLine.substring(beginCharIndex);
+
+					boolean newMethodCall = false;
+					int stack = 0;
```

Comment:
@nicolas-sss, why don't you use the [Stack Data Structure](https://www.javatpoint.com/java-stack) available in Java?

If you decide to use a counter for it, let's name it with a more meaningful name, like `parenthesisStack`